### PR TITLE
Add an argument for scheme pre and post actions in Xcode

### PR DESCRIPTION
**Pull request by robbertvanginkel**
Xcode currently allows you to execute a shell step when you start an stop executing in a scheme. I would like buck to be able to generate a project with this configured.

I have an implementation of this on an internal fork, it works but the interface isn't very satisfying. Although the rule configuration seems reasonable:
```
xcode_workspace_config(
    ...
    additional_scheme_actions = {
      'Build': {
          'Pre': ['echo start'],
          'Post': ['echo stop'],
      }
    }
    ...
)
```
The current argument type that is coerced seems too long
```
Optional<ImmutableMap<SchemeActionType, ImmutableMap<XCScheme.AdditionalActions, ImmutableList<String>>>>
```
Any suggestions on how to handle this? Is this acceptable in java? Should I try to achieve this outside of buck?

### CODE REVIEW

**Comment by ttsugriy** [comment]
@dinhviethoa, do you happen to have some thoughts about this?

File: src/com/facebook/buck/apple/project_generator/SchemeGenerator.java [hunk 1]
```diff
@@ -107,6 +115,7 @@ public SchemeGenerator(
     this.remoteRunnablePath = remoteRunnablePath;
     this.actionConfigNames = actionConfigNames;
     this.targetToProjectPathMap = targetToProjectPathMap;
+    this.additionalSchemeActions = additionalSchemeActions;
```
Comment (dinhviethoa) [inline src/com/facebook/buck/apple/project_generator/SchemeGenerator.java]: You're not using this value to write data to the file yet?

File: src/com/facebook/buck/apple/xcode/XCScheme.java [hunk 2]
```diff
@@ -75,6 +75,12 @@ public String getName() {
     return archiveAction;
   }
 
+  public enum AdditionalActions {
+    PRE,
```
Comment (dinhviethoa) [inline src/com/facebook/buck/apple/xcode/XCScheme.java]: It looks too short :/
Maybe we could use PRE_ACTION nad POST_ACTION.

**Comment by dinhviethoa** [comment]
Is it a work in progress?

**Comment by dinhviethoa** [comment]
About the argument type, it's probably fine as is.

File: src/com/facebook/buck/apple/project_generator/SchemeGenerator.java (hunk 1 above)
Comment (robbertvanginkel) [reply src/com/facebook/buck/apple/project_generator/SchemeGenerator.java]: No not yet, wanted to get feedback first on if the type for `additionalSchemeActions` would be acceptable.

**Comment by robbertvanginkel** [comment]
I have some other commits that make the full implementation, but I wanted to share this first to get feedback on the long argument type. I'll update with the full implementation.

**Comment by robbertvanginkel** [comment]
@dinhviethoa its been a while but I finally found time to rebase this. I incorporated the feedback on the short `AdditionalActions` enum names. It now also includes the serialisation code and appropriate tests.

**Comment by milend** [comment]
LGTM

**Pull request closed without merging**